### PR TITLE
Media library: add data source

### PR DESCRIPTION
**Pull request by johngodley**
Adds a 'source' value to the media library which will be used to determine the data source for the media library - WordPress or something else (Google Photos, for example).

This data source will then be used to determine whether certain bits of UI appear in the media modal (as well as other things in the future).

Note this PR doesn't change the current media modal, and doesn't introduce other data sources.

## Testing

Run included unit tests which verifies that the current media modal shows the same buttons, and that changing the data source will also change the buttons.

Also manually verify that the media library buttons remain the same as before. Likewise for the media menu.

### CODE REVIEW

**Comment by nosolosw** [comment]
I've got this both in local and in calypso.live:

 first-load

**Comment by nosolosw** [comment]
Also, food for thought:

* I can see how the FilterBar could be mostly the same for potentially different sources (WordPress, Google, Dropbox, etc) as long as these sources provide a way to query the data (search button) or metadata for classifying the files (images, docs, videos, audio buttons).
* In the HeaderBar case, though, I guess it will be harder to replicate the actions and, in some cases, may be even necessary not to render it. Instead of using the `source` prop to hide/display buttons, what do you think of injecting different HeaderBar components depending on the `source`?

**Comment by johngodley** [comment]
Ah, the media page has changed since the PR! Previously it was just a mirror of the media library modal, but now it works as an infinite list.

Anyway, I've made the appropriate fixes to have the page displaying correctly.

I've also made a small change to the HeaderBar by creating a new component for external sources. I suspect that when other external sources are added a lot of this will need to change anyway, but that's a task for another day.

**Comment by johngodley** [comment]
Note I additionally disabled the dropzone when viewing external sources as it would make it look like you're uploading to Google Photos.

The ideal solution would be to switch the user back to the WordPress source, but this would require some tricky changes to other parts of the media library and so I think it's something best left for a future version.

**Comment by nosolosw** [comment]
I can see the FilterBar now, but the resize components floats out of the HeaderBar and action buttons are not shown (tried with and without pictures selected). Tested with JetPack and non JetPack websites.

headerbar

**Comment by johngodley** [comment]
Oops, sorry it seems I forgot to include one of the fixes!

Now included and I've double-checked that it works locally and on calypso.live

File: client/components/data/media-list-data/index.jsx [hunk 1]
```diff
@@ -26,6 +26,7 @@ module.exports = React.createClass( {
 
 	propTypes: {
 		siteId: React.PropTypes.number.isRequired,
+		source: React.PropTypes.string,
```
Comment (nosolosw) [inline client/components/data/media-list-data/index.jsx]: The source parameter doesn't seem to be used in this component, neither it has a default value. Is something that you plan to use in upcoming PRs?

File: client/my-sites/media-library/filter-bar.jsx [hunk 2]
```diff
@@ -38,7 +38,8 @@ export class MediaLibraryFilterBar extends Component {
 		basePath: '/media',
 		onFilterChange: noop,
 		onSearch: noop,
-		translate: identity
+		translate: identity,
+		source: '',
```
Comment (nosolosw) [inline client/my-sites/media-library/filter-bar.jsx]: The `source` and `onSourceChange` props haven't been defined in the propTypes array. I understand that, in an upcoming PR, you plan to create a new kind of list component within FilterBar to take care of selecting the different sources, and that component will fire the `onSourceChange`.

File: client/my-sites/media-library/index.jsx [hunk 3]
```diff
@@ -30,6 +30,7 @@ module.exports = React.createClass( {
 		filter: React.PropTypes.string,
 		enabledFilters: React.PropTypes.arrayOf( React.PropTypes.string ),
 		search: React.PropTypes.string,
+		source: React.PropTypes.string,
```
Comment (nosolosw) [inline client/my-sites/media-library/index.jsx]: This component hasn't defined the `onSourceChange` prop.

File: client/post-editor/media-modal/index.jsx [hunk 4]
```diff
@@ -127,6 +127,7 @@ export class EditorMediaModal extends Component {
 		return {
 			filter: '',
 			detailSelectedIndex: 0,
+			source: '',
```
Comment (nosolosw) [inline client/post-editor/media-modal/index.jsx]: Although `source` is defined in the `getDefaultState` utility function, neither `source` or `onSourceChange` are defined as props or have default values.

File: client/components/data/media-list-data/index.jsx (hunk 1 above)
Comment (nosolosw) [inline client/components/data/media-list-data/index.jsx]: Oh! Are the props in this component the ones used by the list-store query? I am not familiar with flux stores and how the query and Flux controller props play together, so I'm going to blindly trust you on this one taking into account that this is a work in progress and it doesn't break anything on my testing :)

**Comment by nosolosw** [comment]
Oh! I forgot to mention one comment about the use of the source variable, for you to consider in later PRs. I've seen that in this PR the valid values for `source` are a void string or `external`. If in the later PRs we still have two states, it might be worth considering to switch this variable to a boolean one and call it something like `isSourceExternal` to improve code readability. In the case that we end up having different values for each source, I'd suggest to avoid the empty string and use some more semantic value.

File: client/my-sites/media-library/filter-bar.jsx (hunk 2 above)
Comment (johngodley) [inline client/my-sites/media-library/filter-bar.jsx]: Yeah, onSourceChange isn't needed here just yet so hasn't been defined.

File: client/post-editor/media-modal/index.jsx (hunk 4 above)
Comment (johngodley) [inline client/post-editor/media-modal/index.jsx]: This is the root for `source` and `onSourceChange`, matching `search` and `filter` - there's nothing passed in.

**Comment by johngodley** [comment]
I agree that in this particular context (with only internal/external data) a boolean would be better. I've flip-flopped many times about how to represent the source before settling on a string that matches how `filter` and `search` work - simple enough right now, but flexible for extra services. Should more services be added I suspect they'll bring enough peculiarities that we'll need to revisit this anyway.

Thanks!

**Pull request merged**